src/picterra/base_client.py

```diff
@@ -0,0 +1,264 @@
+
+    def __iter__(self) -> Iterator[T]:
+        return iter([self._results[i] for i in range(len(self._results))])
+
+    def __str__(self) -> str:
+        return f"{len(self._results)} results from {self._url}"
+
+
+class Feature(TypedDict):
+    type: Literal["Feature"]
+    properties: dict[str, Any]
+    geometry: dict[str, Any]
+
+
+class FeatureCollection(TypedDict):
+    type: Literal["FeatureCollection"]
+    features: list[Feature]
+
+
+class BaseAPIClient:
+    """
+    Base class for Picterra API clients.
+
+    This is subclassed for the different products we have.
+    """
+
+    def __init__(
+        self, api_url: str, timeout: int = 30, max_retries: int = 3, backoff_factor: int = 10
+    ):
+        """
+        Args:
+            api_url: the api's base url. This is different based on the Picterra product used
+                and is typically defined by implementations of this client
+            timeout: number of seconds before the request times out
+            max_retries: max attempts when ecountering gateway issues or throttles; see
+                retry_strategy comment below
+            backoff_factor: factor used nin the backoff algorithm; see retry_strategy comment below
+        """
+        base_url = os.environ.get(
+            "PICTERRA_BASE_URL", "https://app.picterra.ch/"
+        )
+        api_key = os.environ.get("PICTERRA_API_KEY", None)
+        if not api_key:
+            raise APIError("PICTERRA_API_KEY environment variable is not defined")
+        logger.info(
+            "Using base_url=%s, api_url=%s; %d max retries, %d backoff and %s timeout.",
+            base_url,
+            api_url,
+            max_retries,
+            backoff_factor,
+            timeout,
+        )
+        self.base_url = urljoin(base_url, api_url)
+        # Create the session with a default timeout (30 sec), that we can then
+        # override on a per-endpoint basis (will be disabled for file uploads and downloads)
+        self.sess = _RequestsSession(timeout=timeout)
+        # Retry: we set the HTTP codes for our throttle (429) plus possible gateway problems (50*),
+        # and for polling methods (GET), as non-idempotent ones should be addressed via idempotency
+        # key mechanism; given the algorithm is {<backoff_factor> * (2 **<retries-1>}, and we
+        # default to 30s for polling and max 30 req/min, the default 5-10-20 sequence should
+        # provide enough room for recovery
+        retry_strategy = Retry(
+            total=max_retries,
+            status_forcelist=[429, 502, 503, 504],
+            backoff_factor=backoff_factor,
+            allowed_methods=["GET"],
+        )
+        adapter = HTTPAdapter(max_retries=retry_strategy)
+        self.sess.mount("https://", adapter)
+        self.sess.mount("http://", adapter)
+        # Authentication
+        self.sess.headers.update({"X-Api-Key": api_key})
+
+    def _full_url(self, path: str, params: dict[str, Any] | None = None):
```

Comment:
IIUC this will probably break anywhere that the `PICTERRA_BASE_URL` environment variable is being used to override the URL. Probably only affects our nightlies because there's no reason for a customer to override it?

Reply:
I'm not sure I understand what would break? `PICTERRA_BASE_URL` is still read from `BaseAPIClient.__init__`, so if it is set as an override it should be used correctly?

Reply:
So I think we set `PICTERRA_BASE_URL=https://app.picterra.ch/public/api/v2/` (or equivalent for the testing site)

This results in, for example. the list_detectors endpoint calling the following URL:

`https://app.picterra.ch/public/api/v2/public/api/v2/detectors/?page_number=1`

Reply:
I think I'm ok with this, we'll just need to adjust our nightly test config in gitlab to set a base URL of `https://<testing-site>/` rather than including the `public/api/v2/` bit